File: python/fea.py
"""
Structron - 2D plane-frame finite element solver (direct stiffness method).

This is the genuine numerical core: it assembles the global stiffness matrix
from beam-column elements (axial + Euler-Bernoulli bending, 3 DOF/node:
u, v, theta), applies supports and loads (nodal + element distributed),
solves K u = F, and recovers reactions, member end forces and the deflected
shape. Unlike the closed-form formulas it handles arbitrary geometry:
multi-span continuous beams, cantilevers and full portal frames.

Units: N, mm, MPa (so E in MPa, A in mm^2, I in mm^4, lengths in mm).
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field

import numpy as np


@dataclass
class Node:
    x: float
    y: float


@dataclass
class Element:
    n1: int
    n2: int
    E: float          # MPa
    A: float          # mm^2
    I: float          # mm^4 (strong axis)
    Sx: float = 0.0   # mm^3 (section modulus, for stress recovery)
    w: float = 0.0    # transverse distributed load, N/mm (local +y), optional


@dataclass
class FrameModel:
    nodes: list[Node]
    elements: list[Element]
    # node index -> (fix_u, fix_v, fix_theta)
    supports: dict[int, tuple[bool, bool, bool]] = field(default_factory=dict)
    # node index -> (Fx, Fy, M) in N, N, N*mm
    nodal_loads: dict[int, tuple[float, float, float]] = field(default_factory=dict)


def _elem_geom(model: FrameModel, e: Element):
    n1, n2 = model.nodes[e.n1], model.nodes[e.n2]
    dx, dy = n2.x - n1.x, n2.y - n1.y
    L = math.hypot(dx, dy)
    return L, dx / L, dy / L  # length, cos, sin


def _local_stiffness(E, A, I, L):
    a = E * I / L**3
    EA_L = E * A / L
    return np.array([
        [EA_L, 0, 0, -EA_L, 0, 0],
        [0, 12 * a, 6 * a * L, 0, -12 * a, 6 * a * L],
        [0, 6 * a * L, 4 * a * L * L, 0, -6 * a * L, 2 * a * L * L],
        [-EA_L, 0, 0, EA_L, 0, 0],
        [0, -12 * a, -6 * a * L, 0, 12 * a, -6 * a * L],
        [0, 6 * a * L, 2 * a * L * L, 0, -6 * a * L, 4 * a * L * L],
    ])


def _transform(c, s):
    return np.array([
        [c, s, 0, 0, 0, 0],
        [-s, c, 0, 0, 0, 0],
        [0, 0, 1, 0, 0, 0],
        [0, 0, 0, c, s, 0],
        [0, 0, 0, -s, c, 0],
        [0, 0, 0, 0, 0, 1],
    ])


def _fixed_end_local(w, L):
    # transverse UDL w (N/mm, local +y): fixed-end force vector (local DOFs)
    return np.array([0, w * L / 2, w * L * L / 12, 0, w * L / 2, -w * L * L / 12])


def analyze(model: FrameModel, samples_per_elem: int = 21) -> dict:
    """Solve the frame and return displacements, reactions and member results."""
    nn = len(model.nodes)
    ndof = 3 * nn
    K = np.zeros((ndof, ndof))
    F = np.zeros(ndof)

    # nodal loads
    for ni, (fx, fy, m) in model.nodal_loads.items():
        F[3 * ni:3 * ni + 3] += [fx, fy, m]

    elem_cache = []
    for e in model.elements:
        L, c, s = _elem_geom(model, e)
        kl = _local_stiffness(e.E, e.A, e.I, L)
        T = _transform(c, s)
        kg = T.T @ kl @ T
        dofs = [3 * e.n1, 3 * e.n1 + 1, 3 * e.n1 + 2,
                3 * e.n2, 3 * e.n2 + 1, 3 * e.n2 + 2]
        for i in range(6):
            for j in range(6):
                K[dofs[i], dofs[j]] += kg[i, j]
        fef_l = _fixed_end_local(e.w, L) if e.w else np.zeros(6)
        if e.w:
            F[dofs] += -(T.T @ fef_l)  # equivalent nodal loads
        elem_cache.append((e, L, T, kl, fef_l, dofs))

    # boundary conditions
    fixed = np.zeros(ndof, dtype=bool)
    for ni, (fu, fv, ft) in model.supports.items():
        if fu:
            fixed[3 * ni] = True
        if fv:
            fixed[3 * ni + 1] = True
        if ft:
            fixed[3 * ni + 2] = True
    free = ~fixed

    u = np.zeros(ndof)
    u[free] = np.linalg.solve(K[np.ix_(free, free)], F[free])
    reactions = K @ u - F  # nonzero at fixed DOFs

    # member recovery + deflected shape
    members = []
    max_defl = 0.0
    max_moment = 0.0
    max_stress = 0.0
    for e, L, T, kl, fef_l, dofs in elem_cache:
        u_local = T @ u[dofs]
        f_local = kl @ u_local + fef_l  # true member end forces
        m1, m2 = f_local[2], f_local[5]
        # mid-span moment for a member under UDL (parabolic correction)
        m_mid = -(m1 - m2) / 2 + e.w * L * L / 8 if e.w else (m1 - m2) / 2
        m_max_member = max(abs(m1), abs(m2), abs(m_mid))
        max_moment = max(max_moment, m_max_member)
        if e.Sx:
            max_stress = max(max_stress, m_max_member / e.Sx)
        # transverse deflection via Hermite shape functions
        v1, th1, v2, th2 = u_local[1], u_local[2], u_local[4], u_local[5]
        for k in range(samples_per_elem):
            xi = k / (samples_per_elem - 1)
            N1 = 1 - 3 * xi**2 + 2 * xi**3
            N2 = L * (xi - 2 * xi**2 + xi**3)
            N3 = 3 * xi**2 - 2 * xi**3
            N4 = L * (-xi**2 + xi**3)
            v = N1 * v1 + N2 * th1 + N3 * v2 + N4 * th2
            max_defl = max(max_defl, abs(v))
        members.append({
            "n1": e.n1, "n2": e.n2, "axial_n": f_local[0],
            "moment1_nmm": m1, "moment2_nmm": m2, "m_max_nmm": m_max_member,
        })

    return {
        "displacements": u,
        "reactions": reactions,
        "members": members,
        "max_deflection_mm": max_defl,
        "max_moment_nmm": max_moment,
        "max_stress_mpa": max_stress,
        "ndof": ndof,
    }


# --- model builders --------------------------------------------------------

def simply_supported_beam(
    span_mm: float, total_force_n: float, E: float, I: float, A: float,
    Sx: float, n_elems: int = 12, load: str = "udl",
) -> FrameModel:
    """Build a horizontal simply supported beam (pin + roller) for validation."""
    n_nodes = n_elems + 1
    nodes = [Node(span_mm * i / n_elems, 0.0) for i in range(n_nodes)]
    w = total_force_n / span_mm if load == "udl" else 0.0
    elems = [
        Element(i, i + 1, E, A, I, Sx, w=w) for i in range(n_elems)
    ]
    model = FrameModel(nodes=nodes, elements=elems)
    model.supports[0] = (True, True, False)        # pin
    model.supports[n_nodes - 1] = (False, True, False)  # roller
    if load == "point-mid":
        mid = n_nodes // 2
        model.nodal_loads[mid] = (0.0, total_force_n, 0.0)
    return model


def portal_frame(
    span_mm: float, height_mm: float, total_force_n: float,
    E: float, I: float, A: float, Sx: float, n_beam: int = 8,
) -> FrameModel:
    """Two columns + a top beam carrying a UDL; pinned column bases."""
    nodes: list[Node] = []
    base_l = len(nodes); nodes.append(Node(0, 0))            # 0 base left
    top_l = len(nodes); nodes.append(Node(0, height_mm))     # 1 top left
    # beam nodes left->right at top
    beam_nodes = [top_l]
    for i in range(1, n_beam):
        beam_nodes.append(len(nodes)); nodes.append(Node(span_mm * i / n_beam, height_mm))
    top_r = len(nodes); nodes.append(Node(span_mm, height_mm))  # top right
    beam_nodes.append(top_r)
    base_r = len(nodes); nodes.append(Node(span_mm, 0))        # base right

    w = total_force_n / span_mm
    elems = [Element(base_l, top_l, E, A, I, Sx)]             # left column
    for i in range(len(beam_nodes) - 1):
        elems.append(Element(beam_nodes[i], beam_nodes[i + 1], E, A, I, Sx, w=w))
    elems.append(Element(base_r, top_r, E, A, I, Sx))         # right column

    model = FrameModel(nodes=nodes, elements=elems)
    model.supports[base_l] = (True, True, False)   # pinned base
    model.supports[base_r] = (True, True, False)
    return model


if __name__ == "__main__":
    # Validate against the Structron closed-form benchmark.
    E = 200000.0
    I = 119000 * 1e4   # HW 428x407x20x35, mm^4
    A = 361.4 * 1e2
    Sx = 5580 * 1e3
    F = 4320 * 9.81 * 3  # 127,138 N

    for load in ("udl", "point-mid"):
        m = simply_supported_beam(12000, F, E, I, A, Sx, n_elems=12, load=load)
        r = analyze(m)
        print(
            f"{load:>10}: delta_max = {r['max_deflection_mm']:.2f} mm, "
            f"sigma_max = {r['max_stress_mpa']:.2f} MPa"
        )
    print("Closed-form: udl 12.02 mm / 34.18 MPa ; point-mid 19.23 mm / 68.35 MPa")

File: src/fea.ts
/**
 * 2D plane-frame finite element solver (TypeScript port of python/fea.py).
 *
 * Direct stiffness method, beam-column elements (axial + Euler-Bernoulli
 * bending, 3 DOF/node). Runs in the browser so the Frame Studio can solve
 * arbitrary structures - multi-span beams, cantilevers, portal frames -
 * client-side and animate the deflected shape. Units: N, mm, MPa.
 */

export interface FNode {
  x: number;
  y: number;
}
export interface FElement {
  n1: number;
  n2: number;
  E: number;
  A: number; // mm^2
  I: number; // mm^4
  Sx: number; // mm^3
  w?: number; // transverse UDL, N/mm (local +y)
}
export interface FrameModel {
  nodes: FNode[];
  elements: FElement[];
  supports: Record<number, [boolean, boolean, boolean]>; // fix u,v,theta
  loads: Record<number, [number, number, number]>; // Fx,Fy,M
}

export interface FrameResult {
  u: number[];
  ndof: number;
  maxDeflectionMm: number;
  maxStressMpa: number;
  maxMomentNmm: number;
  /** Per-element sampled points, undeformed and deformed (scale applied). */
  undeformed: { x: number; y: number }[][];
  deformedFn: (scale: number) => { x: number; y: number }[][];
}

function geom(m: FrameModel, e: FElement) {
  const a = m.nodes[e.n1];
  const b = m.nodes[e.n2];
  const dx = b.x - a.x;
  const dy = b.y - a.y;
  const L = Math.hypot(dx, dy);
  return { L, c: dx / L, s: dy / L };
}

/**
 * 6x6 stiffness matrix of ONE beam-column element in its local axes.
 * The 6 rows/cols are the element's degrees of freedom: [u1, v1, theta1, u2,
 * v2, theta2] = axial, transverse and rotation at each of the two nodes.
 * `ea` is the axial term E*A/L; `a` groups the bending terms E*I/L^3.
 */
function localK(E: number, A: number, I: number, L: number): number[][] {
  const a = (E * I) / L ** 3; // bending stiffness group
  const ea = (E * A) / L; // axial stiffness
  return [
    [ea, 0, 0, -ea, 0, 0],
    [0, 12 * a, 6 * a * L, 0, -12 * a, 6 * a * L],
    [0, 6 * a * L, 4 * a * L * L, 0, -6 * a * L, 2 * a * L * L],
    [-ea, 0, 0, ea, 0, 0],
    [0, -12 * a, -6 * a * L, 0, 12 * a, -6 * a * L],
    [0, 6 * a * L, 2 * a * L * L, 0, -6 * a * L, 4 * a * L * L],
  ];
}

function transform(c: number, s: number): number[][] {
  return [
    [c, s, 0, 0, 0, 0],
    [-s, c, 0, 0, 0, 0],
    [0, 0, 1, 0, 0, 0],
    [0, 0, 0, c, s, 0],
    [0, 0, 0, -s, c, 0],
    [0, 0, 0, 0, 0, 1],
  ];
}

function matmul(A: number[][], B: number[][]): number[][] {
  const n = A.length;
  const m = B[0].length;
  const k = B.length;
  const out = Array.from({ length: n }, () => new Array(m).fill(0));
  for (let i = 0; i < n; i++)
    for (let j = 0; j < m; j++) {
      let s = 0;
      for (let p = 0; p < k; p++) s += A[i][p] * B[p][j];
      out[i][j] = s;
    }
  return out;
}
function transpose(A: number[][]): number[][] {
  return A[0].map((_, j) => A.map((r) => r[j]));
}
function matvec(A: number[][], v: number[]): number[] {
  return A.map((r) => r.reduce((s, x, j) => s + x * v[j], 0));
}

/** Gaussian elimination with partial pivoting. */
function solveLinear(A: number[][], b: number[]): number[] {
  const n = b.length;
  const M = A.map((r, i) => [...r, b[i]]);
  for (let col = 0; col < n; col++) {
    let piv = col;
    for (let r = col + 1; r < n; r++)
      if (Math.abs(M[r][col]) > Math.abs(M[piv][col])) piv = r;
    [M[col], M[piv]] = [M[piv], M[col]];
    const d = M[col][col] || 1e-12;
    for (let r = 0; r < n; r++) {
      if (r === col) continue;
      const f = M[r][col] / d;
      for (let k = col; k <= n; k++) M[r][k] -= f * M[col][k];
    }
  }
  return M.map((r, i) => r[n] / (M[i][i] || 1e-12));
}

function fixedEndLocal(w: number, L: number): number[] {
  return [0, (w * L) / 2, (w * L * L) / 12, 0, (w * L) / 2, -(w * L * L) / 12];
}

/**
 * The direct-stiffness FE solver. Steps:
 *   1) build the global stiffness matrix K and load vector F,
 *   2) assemble each element's stiffness into K (rotated to global axes),
 *   3) apply the supports (remove fixed DOFs),
 *   4) solve K*u = F for the nodal displacements u,
 *   5) recover member forces, stresses and the deflected shape.
 */
export function analyzeFrame(model: FrameModel): FrameResult {
  const nn = model.nodes.length; // number of nodes
  const ndof = 3 * nn; // total degrees of freedom (3 per node)
  const K = Array.from({ length: ndof }, () => new Array(ndof).fill(0)); // global stiffness K
  const F = new Array(ndof).fill(0); // global load vector F

  for (const [ni, l] of Object.entries(model.loads)) {
    const i = Number(ni) * 3;
    F[i] += l[0];
    F[i + 1] += l[1];
    F[i + 2] += l[2];
  }

  const cache: {
    e: FElement;
    L: number;
    T: number[][];
    kl: number[][];
    fef: number[];
    dofs: number[];
  }[] = [];
  for (const e of model.elements) {
    const { L, c, s } = geom(model, e);
    const kl = localK(e.E, e.A, e.I, L); // element stiffness in local axes
    const T = transform(c, s); // rotation matrix (c, s = cos, sin of the element angle)
    const kg = matmul(matmul(transpose(T), kl), T); // rotate to global axes: kg = T^T * kl * T
    const dofs = [
      3 * e.n1, 3 * e.n1 + 1, 3 * e.n1 + 2, // global DOF indices of node 1
      3 * e.n2, 3 * e.n2 + 1, 3 * e.n2 + 2, // global DOF indices of node 2
    ];
    // scatter-add this element's 6x6 stiffness into the global matrix K
    for (let i = 0; i < 6; i++)
      for (let j = 0; j < 6; j++) K[dofs[i]][dofs[j]] += kg[i][j];
    const fef = e.w ? fixedEndLocal(e.w, L) : new Array(6).fill(0);
    if (e.w) {
      const eq = matvec(transpose(T), fef);
      for (let i = 0; i < 6; i++) F[dofs[i]] -= eq[i];
    }
    cache.push({ e, L, T, kl, fef, dofs });
  }

  const fixed = new Array(ndof).fill(false);
  for (const [ni, sup] of Object.entries(model.supports)) {
    const i = Number(ni) * 3;
    if (sup[0]) fixed[i] = true;
    if (sup[1]) fixed[i + 1] = true;
    if (sup[2]) fixed[i + 2] = true;
  }
  const freeIdx: number[] = [];
  for (let i = 0; i < ndof; i++) if (!fixed[i]) freeIdx.push(i);

  // Reduce the system to the FREE DOFs only (drop the fixed/support rows and
  // columns), then solve Kff * uf = Ff for the unknown free displacements.
  const Kff = freeIdx.map((r) => freeIdx.map((cc) => K[r][cc])); // reduced stiffness
  const Ff = freeIdx.map((r) => F[r]); // reduced load vector
  const uf = solveLinear(Kff, Ff); // <-- the actual FE solve (Gaussian elimination)
  const u = new Array(ndof).fill(0); // full displacement vector (fixed DOFs stay 0)
  freeIdx.forEach((idx, k) => (u[idx] = uf[k]));

  // recovery: stress, moment, and deflected-shape sampler
  let maxStress = 0;
  let maxMoment = 0;
  let maxDefl = 0;
  const NS = 12;
  const undeformed: { x: number; y: number }[][] = [];
  const localData: {
    a: FNode;
    c: number;
    s: number;
    L: number;
    ul: number[];
  }[] = [];
  for (const { e, L, T, kl, fef, dofs } of cache) {
    const ug = dofs.map((d) => u[d]);
    const ul = matvec(T, ug);
    const fl = matvec(kl, ul).map((v, i) => v + fef[i]);
    const mMax = Math.max(Math.abs(fl[2]), Math.abs(fl[5]),
      e.w ? Math.abs(-(fl[2] - fl[5]) / 2 + (e.w * L * L) / 8) : 0);
    maxMoment = Math.max(maxMoment, mMax);
    if (e.Sx) maxStress = Math.max(maxStress, mMax / e.Sx);
    const { c, s } = geom(model, e);
    const a = model.nodes[e.n1];
    localData.push({ a, c, s, L, ul });
    const und: { x: number; y: number }[] = [];
    for (let k = 0; k <= NS; k++) {
      const xi = k / NS;
      und.push({ x: a.x + c * xi * L, y: a.y + s * xi * L });
      const v =
        (1 - 3 * xi ** 2 + 2 * xi ** 3) * ul[1] +
        L * (xi - 2 * xi ** 2 + xi ** 3) * ul[2] +
        (3 * xi ** 2 - 2 * xi ** 3) * ul[4] +
        L * (-(xi ** 2) + xi ** 3) * ul[5];
      maxDefl = Math.max(maxDefl, Math.abs(v));
    }
    undeformed.push(und);
  }

  const deformedFn = (scale: number) =>
    localData.map(({ a, c, s, L, ul }) => {
      const pts: { x: number; y: number }[] = [];
      for (let k = 0; k <= NS; k++) {
        const xi = k / NS;
        const uax = (1 - xi) * ul[0] + xi * ul[3];
        const v =
          (1 - 3 * xi ** 2 + 2 * xi ** 3) * ul[1] +
          L * (xi - 2 * xi ** 2 + xi ** 3) * ul[2] +
          (3 * xi ** 2 - 2 * xi ** 3) * ul[4] +
          L * (-(xi ** 2) + xi ** 3) * ul[5];
        const dxg = c * uax - s * v;
        const dyg = s * uax + c * v;
        pts.push({
          x: a.x + c * xi * L + scale * dxg,
          y: a.y + s * xi * L + scale * dyg,
        });
      }
      return pts;
    });

  return {
    u,
    ndof,
    maxDeflectionMm: maxDefl,
    maxStressMpa: maxStress,
    maxMomentNmm: maxMoment,
    undeformed,
    deformedFn,
  };
}

// --- preset builders -------------------------------------------------------

type Sec = { E: number; A: number; I: number; Sx: number };

/** Simply supported beam at a chosen mesh density (for convergence + ML data). */
export function simpleBeamFE(
  spanMm: number,
  totalForceN: number,
  sec: Sec,
  nElems = 12,
  load: "udl" | "point-mid" = "udl",
): FrameResult {
  const nodes = Array.from({ length: nElems + 1 }, (_, i) => ({
    x: (spanMm * i) / nElems,
    y: 0,
  }));
  const w = load === "udl" ? totalForceN / spanMm : 0;
  const elements = Array.from({ length: nElems }, (_, i) => ({
    n1: i,
    n2: i + 1,
    E: sec.E,
    A: sec.A,
    I: sec.I,
    Sx: sec.Sx,
    w,
  }));
  const supports: Record<number, [boolean, boolean, boolean]> = {
    0: [true, true, false],
    [nElems]: [false, true, false],
  };
  const loads: Record<number, [number, number, number]> =
    load === "point-mid"
      ? { [Math.floor(nElems / 2)]: [0, totalForceN, 0] }
      : {};
  return analyzeFrame({ nodes, elements, supports, loads });
}

export function presetModel(
  kind: "simple" | "cantilever" | "two-span" | "portal",
  spanMm: number,
  heightMm: number,
  totalForceN: number,
  sec: Sec,
): FrameModel {
  const w = totalForceN / spanMm;
  const { E, A, I, Sx } = sec;
  if (kind === "cantilever") {
    const n = 8;
    const nodes = Array.from({ length: n + 1 }, (_, i) => ({
      x: (spanMm * i) / n,
      y: 0,
    }));
    const elements = Array.from({ length: n }, (_, i) => ({
      n1: i, n2: i + 1, E, A, I, Sx, w,
    }));
    return { nodes, elements, supports: { 0: [true, true, true] }, loads: {} };
  }
  if (kind === "two-span") {
    const n = 12;
    const nodes = Array.from({ length: n + 1 }, (_, i) => ({
      x: (spanMm * i) / n,
      y: 0,
    }));
    const elements = Array.from({ length: n }, (_, i) => ({
      n1: i, n2: i + 1, E, A, I, Sx, w,
    }));
    return {
      nodes,
      elements,
      supports: {
        0: [true, true, false],
        [n / 2]: [false, true, false],
        [n]: [false, true, false],
      },
      loads: {},
    };
  }
  if (kind === "portal") {
    const nb = 8;
    const nodes: FNode[] = [{ x: 0, y: 0 }, { x: 0, y: heightMm }];
    const beam = [1];
    for (let i = 1; i < nb; i++) {
      beam.push(nodes.length);
      nodes.push({ x: (spanMm * i) / nb, y: heightMm });
    }
    const topR = nodes.length;
    nodes.push({ x: spanMm, y: heightMm });
    beam.push(topR);
    const baseR = nodes.length;
    nodes.push({ x: spanMm, y: 0 });
    const elements: FElement[] = [{ n1: 0, n2: 1, E, A, I, Sx }];
    for (let i = 0; i < beam.length - 1; i++)
      elements.push({ n1: beam[i], n2: beam[i + 1], E, A, I, Sx, w });
    elements.push({ n1: baseR, n2: topR, E, A, I, Sx });
    return {
      nodes,
      elements,
      supports: { 0: [true, true, false], [baseR]: [true, true, false] },
      loads: {},
    };
  }
  // simple supported
  const n = 12;
  const nodes = Array.from({ length: n + 1 }, (_, i) => ({
    x: (spanMm * i) / n,
    y: 0,
  }));
  const elements = Array.from({ length: n }, (_, i) => ({
    n1: i, n2: i + 1, E, A, I, Sx, w,
  }));
  return {
    nodes,
    elements,
    supports: { 0: [true, true, false], [n]: [false, true, false] },
    loads: {},
  };
}
